util/clustering/kmeans_clustering.py

```diff
@@ -0,0 +1,231 @@
+# Based on the blog post "How to cluster images based on visual similarity" 
+# written by Gabe Flomo.
+
+
+# for loading/processing the images  
+from keras.preprocessing.image import load_img 
+from keras.preprocessing.image import img_to_array 
+from keras.applications.vgg16 import preprocess_input 
+
+# models 
+from keras.applications.vgg16 import VGG16 
+from keras.models import Model
+
+# clustering and dimension reduction
+from sklearn.cluster import KMeans
+from sklearn.decomposition import PCA
+
+# for everything else
+import argparse
+import os
+import numpy as np
+import matplotlib.pyplot as plt
+from PIL import Image
+from tqdm import tqdm
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Clustering')
+    if 'SM_OUTPUT_DATA_DIR' in os.environ: 
+
+        parser.add_argument('--output-data-dir', type=str, default=os.environ['SM_OUTPUT_DATA_DIR'])
+        #parser.add_argument('--model-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+        parser.add_argument('--input-data-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+    else: 
+        parser.add_argument('--output-data-dir', type=str)
+        parser.add_argument('--input-data-dir', type=str)   
+    
+    parser.add_argument('--analyze-clusters', type=int, default=3)   
+
+    args = parser.parse_args()
+
+    
+    file_dir = os.getcwd()
+    # # change the working directory to the path where the images are located
+    output_path = args.input_data_dir
+    os.chdir(output_path)
+
+    detection_train_images = []
```

Comment:
The variable naming isn't terrible, but it will get confusing if we want to use it on other image sets. A better name would be simple 'images' since there are no other images in the entire script.

Reply:
Sorry I didn't finish typing the name: 'image_fname'